{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Introduction\n",
    "---\n",
    "\n",
    "This Jupyter notebook was made with the intent of streamlining generation of PCR Master Mix (MMX) spreadsheets. The specific application this notebook was made in regards to is running DS PCR reactions on library samples before submission to Hartwell for sequencing.\n",
    "\n",
    "# Instructions for This Notebook\n",
    "---\n",
    "\n",
    "1. Open this file in a Jupyter notebook editor of your choice (Python 3.7.7+).\n",
    "\n",
    "2. Change user variables below:\n",
    "\n",
    "    * LIB_AND_PRIM_DIR: Should stay the same each time. Meant to point at the directory housing the spreadsheet referenced by LIB_AND_PRIM_XLSX.\n",
    "\n",
    "    * LIB_AND_PRIM_XLSX: Should stay same each time. Title of spreadsheet file that contains information on all of the vectors (and their associated primer sets) currently in use in our library workflows.\n",
    "\n",
    "    * INPUT_TSV: Can be changed if so desired, but will need to rename file as well. File in current directory that has information on the samples you would like to include. Tab-separated values file ('\\t') with four columns: sample_name, vector, ng/uL, and uL.\n",
    "\n",
    "        * sample_name: sample name you would like used for the sample on resultant files. Do not include primer name on this, as it will be added later.\n",
    "\n",
    "        * vector: Vector used/containing your library, to be used for matching to associated primer sets.\n",
    "\n",
    "        * ng/uL: Current or reported concentration of DNA in sample, as reported in ng per microliter\n",
    "\n",
    "        * uL: Current volume of sample, as reported in uL.\n",
    "\n",
    "    * SIZE_OF_LIBRARY: Number of gRNAs included in the library being processed\n",
    "\n",
    "    * COVERAGE: Number of replicates per gene that has been deemed necessary for representative coverage of the library.\n",
    "\n",
    "    * EXTRA_RXNS: Extra reactions to account for in volume calculations to account for negative control as well as pipetting errors.\n",
    "\n",
    "3. Edit the input file (referenced and described as INPUT_TSV above) to reflect your current/desired samples.\n",
    "\n",
    "4. Run all cells in this notebook.\n",
    "\n",
    "**Process complete!**\n",
    "\n",
    "Explanation of file output may be found at the bottom of this notebook. After running the notebook, see the \"Follow Up Steps\" section after \"Output\" section to find next experimental steps to perform after generation of the spreadsheets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "import os, sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from IPython.display import display, HTML\n",
    "import openpyxl\n",
    "import datetime\n",
    "from shutil import copy\n",
    "from localModules import combine_mmx_output\n",
    "\n",
    "FILE_SUFFIX = 'tecan_test_pmh'# Goes at end of all filenames\n",
    "EXTRA_RXNS = 1 # 3 allows one repeat -- 1 allows skipping mmx transfer\n",
    "\n",
    "CUR_DIR = os.path.abspath('')\n",
    "LIB_AND_PRIM_DIR = os.path.abspath(r\"Z:\\ResearchHome\\Groups\\millergrp\\home\\common\\Screens\")\n",
    "LIB_AND_PRIM_XLSX = os.path.join(LIB_AND_PRIM_DIR, 'lib_vectors_and_primers.xlsx')\n",
    "\n",
    "INPUT_TSV = os.path.join(CUR_DIR, 'input_pcrmmx_v2.tsv')\n",
    "\n",
    "UPDATING_PRIMERS = True\n",
    "FIRST_PRIMER_TO_USE = {\n",
    "    'pLentiCRISPRv2': 'PC448',\n",
    "    'pLentiGuide-Puro': 'PC92',\n",
    "    'PC529_LVA': 'PC529-B06',\n",
    "    'CROPseq': 'JS26.A01.DS.R8'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Most recent primers for each vector set:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>most_recent_used</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>primer_set</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>pLentiCRISPRv2</th>\n",
       "      <td>PC474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pLentiGuide-Puro</th>\n",
       "      <td>PC396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PC529_LVA</th>\n",
       "      <td>PC529-B12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CROPseq</th>\n",
       "      <td>JS26.A01.DS.R8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 most_recent_used\n",
       "primer_set                       \n",
       "pLentiCRISPRv2              PC474\n",
       "pLentiGuide-Puro            PC396\n",
       "PC529_LVA               PC529-B12\n",
       "CROPseq            JS26.A01.DS.R8"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "vectors_df = pd.read_excel(LIB_AND_PRIM_XLSX, sheet_name='vectors',engine='openpyxl')\n",
    "primers_df = pd.read_excel(LIB_AND_PRIM_XLSX, sheet_name='primers_for_screening',engine='openpyxl')\n",
    "mrp_df = pd.read_excel(LIB_AND_PRIM_XLSX, sheet_name='most_recent_primer', index_col=0,engine='openpyxl') # \"Most recent Primers\" df\n",
    "\n",
    "input_df = pd.read_csv(INPUT_TSV, sep='\\t')\n",
    "\n",
    "print('Most recent primers for each vector set:')\n",
    "display(mrp_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     CAGE#                                      Name               Digest  \\\n",
      "0   SNP112                        pLentiCRISPRv3-GFP         Esp3I/ BsmBI   \n",
      "1    SNP36                       pLentiCRISPRv2-Puro         Esp3I/ BsmBI   \n",
      "2    SNP28                           pLentiGuidePuro         Esp3I/ BsmBI   \n",
      "3     SP16                          pLentiGuide-eGFP         Esp3I/ BsmBI   \n",
      "4    PC291  pXPR_502_lentiguide-P65-HSF (Activation)         Esp3I/ BsmBI   \n",
      "5    PC292          pXPR_modified_TRACR (Inhibition)         Esp3I/ BsmBI   \n",
      "6    PC529                              LVA-Ametrine         Esp3I/ BsmBI   \n",
      "7    PC290                        pLentiCRISPRv3-GFP         Esp3I/ BsmBI   \n",
      "8    PC592                              pCRISPRia-v2           Bstx1/Blp1   \n",
      "9    PC226                   LMA-Ametrine_retrovirus                 BbsI   \n",
      "10     NaN                       pLentiCRISPR_v1 (!)                  NaN   \n",
      "11   PC329                                       NaN         Esp3I/ BsmBI   \n",
      "12    JS22                      LMA-DC (from Hongbo)  BstXI/BlpI (or Bup)   \n",
      "13    JS26                        sgOpti_CROPseq_GFP         Esp3I/ BsmBI   \n",
      "14    JS27                             CROP-Seq-opti         Esp3I/ BsmBI   \n",
      "15  SNP338                                  Bip_Papi                  NaN   \n",
      "16   PC632                        CROPseq-Guide-Puro         Esp3I/ BsmBI   \n",
      "17     LMG                       Lentiviral-PGK-mGFP                 BbsI   \n",
      "18     NaN             lenti sgRNA(MS2)_zeo backbone                  NaN   \n",
      "19   PC658                      AsCas12as_Dual_PC658                  NaN   \n",
      "20   MF3-A       pLentiCRISPRv2(Modified)-dCas9-EGFP                BsmBI   \n",
      "\n",
      "       Selection                                    Primers  \\\n",
      "0           eGFP                             pLentiCRISPRv2   \n",
      "1           Puro                             pLentiCRISPRv2   \n",
      "2           Puro                           pLentiGuide-Puro   \n",
      "3           eGFP                           pLentiGuide-Puro   \n",
      "4           Puro                           pLentiGuide-Puro   \n",
      "5           Puro                           pLentiGuide-Puro   \n",
      "6       Ametrine                                  PC529_LVA   \n",
      "7           eGFP                             pLentiCRISPRv2   \n",
      "8   BFP and Puro                               pCRISPRia-v2   \n",
      "9       Ametrine                             pLentiCRISPRv2   \n",
      "10          Puro                           pLentiGuide-Puro   \n",
      "11       mCherry                             pLentiCRISPRv2   \n",
      "12          Puro  Forward JS22:  Reverse: JS22.DS.R I think   \n",
      "13           GFP                                    CROPseq   \n",
      "14          Puro                                    CROPseq   \n",
      "15          Puro                             pLentiCRISPRv2   \n",
      "16          Puro                                    CROPseq   \n",
      "17           GFP                             pLentiCRISPRv2   \n",
      "18          test                              LentiCRISPRv2   \n",
      "19           NaN                           pLentiGuide-Puro   \n",
      "20          eGFP                                  PC529_LVA   \n",
      "\n",
      "                                                Notes  \\\n",
      "0                                                 NaN   \n",
      "1                                                 NaN   \n",
      "2                                                 NaN   \n",
      "3                                                 NaN   \n",
      "4                                                 NaN   \n",
      "5                                                 NaN   \n",
      "6                                    AKA: LVA, LVA-V4   \n",
      "7                                 Made by vector core   \n",
      "8                         Weissman library CRISPRa-v2   \n",
      "9                                                 NaN   \n",
      "10  Yes, LentiPuro primers even though it's LentiC...   \n",
      "11  NOT SURE ABOUT INFORMATION -- DOUBLE CHECK BEF...   \n",
      "12                                                NaN   \n",
      "13                                                NaN   \n",
      "14                                                NaN   \n",
      "15                NGS requires Single end 300bp reads   \n",
      "16                                                NaN   \n",
      "17                                    From Hongbo lab   \n",
      "18                        PC84-91 and PC94 primer set   \n",
      "19                                                NaN   \n",
      "20  Fragmid vector - pLentiCRISPRv2(Modified)-dCas...   \n",
      "\n",
      "                                     SRM Name Addgene_num  \n",
      "0        ​LentiCRISPR.v2-eGFP (gRNA and Cas9)      ​82416  \n",
      "1        ​LentiCRISPR.v2-Puro (gRNA and Cas9)      ​52961  \n",
      "2                  LentiGuide-Puro (w/o Cas9)       52963  \n",
      "3                ​LentiGuide- eGFP (w/o Cas9)      ​65656  \n",
      "4             ​pXPR_502 p65-HSF (for CRISPRa)      96923​  \n",
      "5   ​pXPR_050 (modified tracrRNA for CRISPRi)      ​96925  \n",
      "6              LentiGuide-Ametrine (w/o Cas9)         NaN  \n",
      "7                                         NaN         NaN  \n",
      "8                                         NaN         NaN  \n",
      "9                                         NaN         NaN  \n",
      "10                                        NaN         NaN  \n",
      "11    ​LentiCRISPR.v2-mCherry (gRNA and Cas9)      ​99154  \n",
      "12                                        NaN         NaN  \n",
      "13                                        NaN         NaN  \n",
      "14                                        NaN      106280  \n",
      "15                                        NaN         NaN  \n",
      "16                                        NaN         NaN  \n",
      "17                                        NaN         NaN  \n",
      "18                                        NaN         NaN  \n",
      "19                                        NaN         NaN  \n",
      "20                                        NaN         NaN  \n",
      "PC658 requires 2 primers. (Primer set: pLentiGuide-Puro)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "vec_to_set = {} # Use this at some point to convert to handling multiple vectors/sets at once\n",
    "expmt_primers = {}\n",
    "input_df['primer_set'] = ''\n",
    "# If actually using samples from more than one vector at a time, will need to write primers to input_df accordingly\n",
    "for vector in input_df['vector'].unique():\n",
    "    count = len(input_df[input_df['vector'] == vector].index)\n",
    "    print(vectors_df)\n",
    "    vector_row = vectors_df.index[vectors_df['CAGE#'] == vector][0]\n",
    "    # print(vector_row)\n",
    "    primers_col = vectors_df.columns.get_loc('Primers')\n",
    "    primer_set = vectors_df.iat[vector_row, primers_col]\n",
    "    vec_to_set[vector] = primer_set\n",
    "    print(f'{vector} requires {count} primers. (Primer set: {primer_set})\\n')\n",
    "    try:\n",
    "        expmt_primers[primer_set] += count\n",
    "    except KeyError:\n",
    "        expmt_primers[primer_set] = count\n",
    "    except Exception as e:\n",
    "        print(e)\n",
    "    \n",
    "    input_df.loc[input_df['vector'] == vector, 'primer_set'] = primer_set\n",
    "\n",
    "# print(vec_to_set)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'set': 'pLentiGuide-Puro', 'series': 46    PC397\n",
      "47    PC398\n",
      "Name: primer_name, dtype: object}]\n"
     ]
    }
   ],
   "source": [
    "# Iterate through vectors found in this experiment's input df and find primers for each\n",
    "# To actually implement for multiple vectors, will need to assign primers to rows properly\n",
    "#   in the non-sliced input_df\n",
    "\n",
    "final_list = []\n",
    "final_primer_df = pd.DataFrame()\n",
    "\n",
    "for set, count in expmt_primers.items():\n",
    "    vset, vcount = (set, count)\n",
    "    vector_primers = primers_df[primers_df['primer_group'] == vset]\n",
    "    #print(vector_primers)\n",
    "    recent_primer = mrp_df.at[vset, 'most_recent_used']\n",
    "    if not UPDATING_PRIMERS:\n",
    "        recent_primer = FIRST_PRIMER_TO_USE[vset]\n",
    "    recent_index = vector_primers.reset_index().set_index('primer_name').loc[recent_primer, 'level_0']\n",
    "    #print(recent_index)\n",
    "    primer_df_indices = vector_primers.index.to_list()\n",
    "    loc_in_primer_df = primer_df_indices.index(recent_index)\n",
    "    #print(loc_in_primer_df)\n",
    "    #print(primer_df_indices[loc_in_primer_df])\n",
    "    shifted_indices = primer_df_indices[loc_in_primer_df + 1:] + primer_df_indices[:loc_in_primer_df]\n",
    "    shifted_df = vector_primers.loc[shifted_indices]\n",
    "    # print(shifted_df)\n",
    "    vprimers_series = shifted_df['primer_name'].head(vcount)\n",
    "    final_list.append({'set': vset, 'series': vprimers_series})\n",
    "    \n",
    "    print(final_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# Begin generating output DataFrame\n",
    "\n",
    "output_df = input_df.copy(deep=True)\n",
    "\n",
    "# output_df['primer'] = final_list[0]['list'] # NOTE: This relies on only one vector/primer set being used.\n",
    "output_df['primer'] = ''\n",
    "for v in final_list:\n",
    "    output_df.loc[output_df['primer_set'] == v['set'], ['primer']] = v['series'].values\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "def dna_to_screen(row):\n",
    "    return ((row['num_grnas'] * row['coverage']) * (6/1000) / 1000)\n",
    "\n",
    "output_df['ug_to_screen'] = output_df.apply(dna_to_screen, axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# TODO: Maybe fix the hartwell sample name to be a formula instead of hard-coded\n",
    "# As-is, if you make any changes to the sample names or primers, it will not be reflected in the hart_sample_name\n",
    "# output_df['hart_sample_name'] = output_df['sample_name'].map(str) + '_' + output_df['primer'].map(str)\n",
    "\n",
    "output_df['total_ug_DNA'] = round(((output_df['ng/uL'] * output_df['uL']) / 1000), 1) # Total ug of DNA for each sample\n",
    "\n",
    "def uL_for_10ug(ng_per_uL):\n",
    "    results = round(min(10000/ng_per_uL, 25.0), 1)\n",
    "    return results\n",
    "\n",
    "# def enough_dna(total_ug_dna):\n",
    "#     print(total_ug_dna)\n",
    "#     return total_ug_dna > DNA_TO_SCREEN\n",
    "# output_df['total_ug_DNA'].apply(enough_dna)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "\n",
    "output_df['uL DNA/rxn'] = output_df['ng/uL'].map(uL_for_10ug) # uL of sample/DNA per reaction\n",
    "output_df['uL MMX/rxn'] = 75 # 100 - output_df['uL DNA/rxn']\n",
    "output_df['ug/rxn'] = round((output_df['ng/uL'] * output_df['uL DNA/rxn']) / 1000, 2) # ug of DNA loaded per rxn for each sample\n",
    "output_df['total_ug_DNA'] = round(((output_df['ng/uL'] * output_df['uL']) / 1000), 1) # Total ug of DNA for each sample\n",
    "\n",
    "output_df['limit_volume'] = pd.Series(output_df['uL'] / output_df['uL DNA/rxn']).map(np.floor) # Max rxns as limited by volume\n",
    "output_df['limit_DNA'] = pd.Series(output_df['total_ug_DNA'] / output_df['ug/rxn']).map(np.floor) # Max rxns as limited by DNA\n",
    "output_df['required_DNA'] = pd.Series(output_df['ug_to_screen'] / output_df['ug/rxn']).map(np.ceil) # Minimum rxns required to reach desired coverage\n",
    "\n",
    "output_df['n'] = pd.Series(output_df[['limit_volume', 'limit_DNA', 'required_DNA']].min(axis=1)).map(np.ceil) # Finds minimum between max by vol, max by ug DNA, required DNA\n",
    "output_df['ug_screened'] = pd.Series(output_df['ug/rxn'] * output_df['n']) # Total ug of DNA screened\n",
    "output_df['n_total'] = output_df['n'].map(int) + EXTRA_RXNS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\phall3\\Anaconda3\\envs\\cage37\\lib\\site-packages\\ipykernel_launcher.py:3: DeprecationWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.\n",
      "  This is separate from the ipykernel package so we can avoid doing imports until\n"
     ]
    }
   ],
   "source": [
    "# Create Master Mix (MMX) series for 1x Rxn (mmx_per_rxn) and for all rxns (gen_mmx)\n",
    "\n",
    "mmx_per_rxn = pd.Series()\n",
    "mmx_per_rxn['Rxn Volume'] = 100\n",
    "mmx_per_rxn['dNTPs (2.5mM)'] = 8\n",
    "mmx_per_rxn['DMSO'] = 5\n",
    "mmx_per_rxn['10X Titanium Taq Buffer'] = mmx_per_rxn['Rxn Volume'] / 10\n",
    "mmx_per_rxn['PC84-PC91 (10uM)'] = 5\n",
    "mmx_per_rxn['Unique Reverse Primer (10uM)'] = 5\n",
    "mmx_per_rxn['Titanium Taq'] = 1.5\n",
    "\n",
    "# max_sample_volume = output_df['uL DNA/rxn'].max()\n",
    "# mmx_per_rxn['h2o'] = mmx_per_rxn['Rxn Volume'] - (mmx_per_rxn.iloc[1:].sum() + max_sample_volume)\n",
    "mmx_per_rxn['h2o'] = 40.5 # Assuming all rxns will need 25 uL of DNA+H2O\n",
    "# display(big_master_mix)\n",
    "\n",
    "total_rxns = output_df['n_total'].sum()\n",
    "safe_total = total_rxns * 1.05\n",
    "# print(total_rxns)\n",
    "# print(safe_total)\n",
    "\n",
    "gen_mmx = mmx_per_rxn[1:] * safe_total\n",
    "gen_mmx = gen_mmx.drop('Unique Reverse Primer (10uM)')\n",
    "gen_mmx['uL of MMX/rxn'] = gen_mmx.sum() / safe_total\n",
    "gen_mmx['Total Volume'] = gen_mmx[~gen_mmx.index.isin(['uL of MMX/rxn'])].sum()\n",
    "# display(big_master_mix_final)\n",
    "\n",
    "output_df['uL MMX for Individual MMX'] = gen_mmx['uL of MMX/rxn'] * output_df['n_total']\n",
    "\n",
    "def water_over_min(uL_DNA_per_rxn):\n",
    "    h2o_to_add_per_rxn = 25 - uL_DNA_per_rxn\n",
    "    return h2o_to_add_per_rxn\n",
    "\n",
    "output_df['uL H2O/rxn'] = output_df['uL DNA/rxn'].map(water_over_min) # * output_df['n_total'] # Uncomment previous multiplication to get water for mmx\n",
    "output_df['uL Rev Primer for individual MMX'] = output_df['n_total'] * mmx_per_rxn['Unique Reverse Primer (10uM)']\n",
    "\n",
    "# TODO: Make two columns below only run for reactions where n > 1 (since these are already covered by original test PCRs)\n",
    "# output_df['uL to Remove'] = (EXTRA_RXNS - 2) * (100 - output_df['uL DNA/rxn']) # Reasoning for this calculation is described in Appendix A\n",
    "output_df['desired MMX vol'] = (output_df['n'] - 1) * 75 # * (100 - output_df['uL DNA/rxn']) # Based on Appendix A -- Desired rxns in MMX * uL DNA/rxn in MMX\n",
    "\n",
    "output_df['uL DNA to Add'] = (output_df['n'] - 1) * output_df['uL DNA/rxn']\n",
    "output_df['uL H2O to Add'] = (output_df['n'] - 1) * output_df['uL H2O/rxn']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# Determine wells to load samples in\n",
    "# according to Hartwell's preference of A01, B01, C01... F12, G12, H12\n",
    "\n",
    "well_rows = 'ABCDEFGH'\n",
    "well_cols = [\n",
    "    f'{c + 1:02d}' for c in range(12)\n",
    "]\n",
    "combs = [r + c for c in well_cols for r in well_rows]\n",
    "\n",
    "output_df['hart_well'] = combs[:len(output_df)]\n",
    "\n",
    "# print(well_rows)\n",
    "# print(well_cols)\n",
    "# print(combs)\n",
    "# display(output_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Today is: 2025_01_17\n",
      "\n",
      "Successfully backed up lib_vectors_and_primers and updated MRP sheet.\n",
      "Most recent primer for set  [ pLentiGuide-Puro ]  has changed.\n",
      "    FROM:  [ PC396 ]\n",
      "      TO:  [ PC398 ]\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'set': 'pLentiGuide-Puro',\n",
       "  'series': 46    PC397\n",
       "  47    PC398\n",
       "  Name: primer_name, dtype: object}]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully saved input DataFrame to TSV.\n",
      "Successfully saved Gen MMX to TSV.\n",
      "Successfully saved full DataFrame to TSV.\n",
      "Successfully saved simplified DataFrame to TSV.\n",
      "\n",
      "Output directory is:\n",
      "    z:\\ResearchHome\\Groups\\millergrp\\projects\\CrisprDesigns\\common\\screens\\programs\\pcr_mmx\\output\\2025_01_17_Lib122_Screen_Mack_841671\n"
     ]
    }
   ],
   "source": [
    "\n",
    "today = datetime.date.today().strftime('%Y_%m_%d')\n",
    "print(f'Today is: {today}\\n')\n",
    "\n",
    "# Make the output directory.\n",
    "output_dir = os.path.join(CUR_DIR, 'output', f'{today}_{FILE_SUFFIX}')\n",
    "try:\n",
    "    if not os.path.exists(output_dir):\n",
    "        os.makedirs(output_dir)\n",
    "except Exception as err:\n",
    "    print(err)\n",
    "\n",
    "# Back up old lib_vectors_and_primers file, update file with new most-recent primers\n",
    "try:\n",
    "    copy(LIB_AND_PRIM_XLSX, os.path.join(LIB_AND_PRIM_DIR, 'archive', 'lib_vectors_and_primers', f'{today}.xlsx'))\n",
    "    print('Successfully backed up lib_vectors_and_primers and updated MRP sheet.')\n",
    "    \n",
    "    integer_ind_mrp_df = mrp_df.reset_index() # Resets the mrp_df index to be integers instead of the primer set names.\n",
    "    \n",
    "    wb = openpyxl.load_workbook(LIB_AND_PRIM_XLSX)\n",
    "    ws = wb['most_recent_primer']\n",
    "    \n",
    "    for set_object in final_list:\n",
    "        new_most_recent = set_object['series'].iat[-1]\n",
    "        #display(new_most_recent, old_most_recent)\n",
    "        old_most_recent = mrp_df.at[set_object['set'], 'most_recent_used']\n",
    "        if not UPDATING_PRIMERS:\n",
    "            new_most_recent = old_most_recent\n",
    "        mrp_index_to_add = integer_ind_mrp_df.loc[\n",
    "            integer_ind_mrp_df['primer_set'] == set_object['set'] # Find row relating to current primer set\n",
    "            ].index.to_list()[0] \n",
    "        #display(f\"mrp_index_to_add {mrp_index_to_add}\")# Get the index as a list, then return first (only) element of the list.\n",
    "        cell = 'B' + str(2 + mrp_index_to_add) # First data cell is B2. Add to row the 0-indexed position in mrp_df.\n",
    "        ws[cell] = new_most_recent\n",
    "        new_prime = cell\n",
    "        #display(new_prime)\n",
    "        \n",
    "        print(f'Most recent primer for set  [ {set_object[\"set\"]} ]  has changed.')\n",
    "        print(f'    FROM:  [ {old_most_recent} ]')\n",
    "        print(f'      TO:  [ {new_most_recent} ]\\n')\n",
    "    wb.save(LIB_AND_PRIM_XLSX)\n",
    "    wb.close()\n",
    "    display(final_list)\n",
    "except Exception as err:\n",
    "    print('\\n\\n!!! Failed to back up lib_vectors_and_primers! Not saving over !!!\\n')\n",
    "    print(err)\n",
    "    print('\\n')\n",
    "    print('For manual entry:')\n",
    "    print(f'    Most recent primer for set  [ {final_list[0][\"set\"]} ]  should change...')\n",
    "    print(f'        FROM:  [ {old_most_recent} ]')\n",
    "    print(f'          TO:  [ {new_most_recent} ]\\n\\n')\n",
    "\n",
    "# Output the input file for logging\n",
    "try:\n",
    "    input_df.to_csv(os.path.join(output_dir, f'{today}_lib_analysis_input.tsv'), sep='\\t', index=False)\n",
    "    print('Successfully saved input DataFrame to TSV.')\n",
    "except Exception as err:\n",
    "    print('\\n\\n!!! Failed to save Input DF to TSV !!!\\n')\n",
    "    print(err)\n",
    "    print('\\n')\n",
    "\n",
    "# Output the generic master mix table\n",
    "try:\n",
    "    gen_mmx.to_csv(os.path.join(output_dir, f'{today}_lib_analysis_mmx_gen.tsv'), sep='\\t', index=True, header=True)\n",
    "    print('Successfully saved Gen MMX to TSV.')\n",
    "except Exception as err:\n",
    "    print('\\n\\n!!! Failed to save Gen MMX to TSV !!!\\n')\n",
    "    print(err)\n",
    "    print('\\n')\n",
    "\n",
    "# Output the full dataframe table\n",
    "try:\n",
    "    output_df.to_csv(os.path.join(output_dir, f'{today}_lib_analysis_mmx_full.tsv'), sep='\\t', index=False)\n",
    "    print('Successfully saved full DataFrame to TSV.')\n",
    "except Exception as err:\n",
    "    print('\\n\\n!!! Failed to save main DataFrame to TSV !!!\\n')\n",
    "    print(err)\n",
    "    print('\\n')\n",
    "\n",
    "# Output the simplified DF table\n",
    "simplified_output_df = output_df[[\n",
    "    'sample_name',\n",
    "    'n_total',\n",
    "    'uL MMX for Individual MMX',\n",
    "    'primer',\n",
    "    'uL Rev Primer for individual MMX',\n",
    "    # 'uL MMX/rxn',\n",
    "    'uL DNA/rxn',\n",
    "    'uL H2O/rxn',\n",
    "    # 'uL to Remove',\n",
    "    'desired MMX vol',\n",
    "    'uL H2O to Add',\n",
    "    'uL DNA to Add',\n",
    "    'hart_well',\n",
    "    ]]\n",
    "try:\n",
    "    simplified_output_df.to_csv(os.path.join(output_dir, f'{today}_lib_analysis_mmx_simplified.tsv'), sep='\\t', index=False)\n",
    "    print('Successfully saved simplified DataFrame to TSV.')\n",
    "except Exception as err:\n",
    "    print('\\n\\n!!! Failed to save main DataFrame to TSV !!!\\n')\n",
    "    print(err)\n",
    "    print('\\n')\n",
    "\n",
    "print(f'\\nOutput directory is:\\n    {output_dir}')\n",
    "\n",
    "# TODO: Ask if they want you to copy somewhere, and take a Windows path as input\n",
    "#   Try/Except copying output_dir/today to desired destination"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "list_of_names = [\"Input\", \"Full Table\", \"Generic MMX\", \"Simplified Table\"]\n",
    "list_of_dfs = [input_df, output_df, gen_mmx, simplified_output_df]\n",
    "list_of_sheets = [\"input\", \"pcr_mmx\", \"gen_mmx\", \"pcr_mmx\"]\n",
    "list_of_calc_cols = [[], [2, 3], [1], [2]]\n",
    "\n",
    "mmx_obj_list = combine_mmx_output.make_list_of_MMXOutputFiles(\n",
    "    list_of_names, list_of_dfs, list_of_sheets, list_of_calc_cols)\n",
    "\n",
    "combine_mmx_output.make_excel(mmx_obj_list, os.path.join(output_dir, f'{today}_mmx-{FILE_SUFFIX}.xlsx'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# Reporting\n",
    "\n",
    "# TODO: Generate a short summary file containing relevant information\n",
    "\n",
    "# print('\\nGen MMX:')\n",
    "# display(gen_mmx)\n",
    "\n",
    "# print('\\nFinal volumes:')\n",
    "# display(output_df[['sample_name', 'hart_sample_name', 'n_total', 'uL MMX for Individual MMX', 'uL H2O for individual Master Mix', 'uL Rev Primer for individual MMX']])"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Output\n",
    "---\n",
    "\n",
    "Expected output of this file will be two TSV files containing the original columns from the input TSV, as well as the following additional columns:\n",
    "\n",
    "* primer: Generated primer choice for sample\n",
    "\n",
    "* hart_sample_name: Sample name to be used in submission to Hartwell\n",
    "\n",
    "* uL DNA/rxn: uL of DNA that will be used per PCR reaction\n",
    "\n",
    "* ug/rxn: ug of DNA that will be used per PCR reaction\n",
    "\n",
    "* total_ug_DNA: Sum total of ug DNA that will be processed by all PCR reactions associated with the sample (n)\n",
    "\n",
    "* limit_volume: Maximum reactions that may be generated, as limited by the **volume** of sample available\n",
    "\n",
    "* limit_DNA: Maximum reactions that may be generated, as limited by the **DNA** of sample available\n",
    "\n",
    "* n: Number of reactions to be made from the sample's individual master mix\n",
    "\n",
    "    * Determined by finding the lowest number out of: maximum by volume, maximum by DNA, minimum to process appropriate ug of DNA for representative analysis\n",
    "\n",
    "* n_total: Total number of reactions that may be fulfilled by the individual master mix (n + EXTRA_RXNS)\n",
    "\n",
    "* uL MMX for Individual MMX: uL to transfer from the large/generic master mix into a new tube for the sample's individual master mix\n",
    "\n",
    "* uL H2O for individual Master Mix: uL of H2O to add to the individual master mix before aliquoting into reactions\n",
    "\n",
    "* uL Rev Primer for individual MMX: uL of individual's designated reverse primer to add to individual MMX before aliquoting into reactions\n",
    "\n",
    "* uL to Remove: uL of individual MMX to remove after rxn confirmation\n",
    "    \n",
    "    * **WARNING!** Only remove this volume **AFTER** confirming test reactions and **BEFORE** adding DNA to the MMX.\n",
    "\n",
    "* uL DNA to Add: After confirming test rxns and removing the uL of MMX from above, add this many uL of DNA to the individual MMX to allow equal mixing and distribution to PCR reactions by a single dispense rather than dispensing DNA to each well, then dispensing MMX to each well.\n",
    "    \n",
    "    * **WARNING!** Only add uL of DNA **AFTER** reaction confirmation by gel and **AFTER** removing the \"uL to Remove\" from individual MMX\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Follow-Up Steps\n",
    "---\n",
    "\n",
    "After running this notebook, use the newly-generated spreadsheets to follow the steps below.\n",
    "\n",
    "\n",
    "\n",
    "## Prep and Label Tubes\n",
    "\n",
    "Begin by getting all of the tubes you will need ready for individual master mixes. Label as necessary.\n",
    "\n",
    "> ### ***Pro Tip!***\n",
    "\n",
    ">> 1.5mL Tubes work for n_total up to 15, 15mL up to 150.\n",
    "\n",
    ">> ...I hope for your sake you don't have more reactions to perform for a sample than 150.\n",
    "\n",
    "## Generic Master Mix\n",
    "\n",
    "1. Use the bigMMX table to determine volumes of reagents to retrieve/begin thawing.\n",
    "\n",
    "2. Prepare and label a tube for the generic MMX with a max volume greater than \"Total Volume\" in the bigMMX table.\n",
    "\n",
    "    * Note: \"uL of MMX/rxn\" is mostly just for reference; It is the volume of Generic MMX that will be present in each individual PCR reaction.\n",
    "\n",
    "3. Add H2O to the tube, then proceed through remaining ingredients in order.\n",
    "\n",
    "\n",
    "\n",
    "## Making individual MMXs\n",
    "\n",
    "1. Add Generic MMX for each sample to the individual MMX tubes in the volume listed in the \"uL MMX for Individual MMX\" column in the mmx table.\n",
    "\n",
    "2. Add H2O according to \"uL H2O for Individual MMX\" column to the individual MMX.\n",
    "\n",
    "3. Add the selected reverse primer for the sample in the volume listed as \"uL Reverse Primer for Individual MMX\"\n",
    "\n",
    "\n",
    "\n",
    "## Running Test Reactions\n",
    "\n",
    "1. Aliquot one reaction for the negative control and one for the positive control into PCR tubes.\n",
    "\n",
    "    1.1. This volume will be $[Reaction Volume] - [uL DNA/rxn]$, so typically 100 uL minus however many uL of DNA you add.\n",
    "\n",
    "2. Add H2O to the negative control and DNA to the positive control to bring the reactions to 100 uL.\n",
    "\n",
    "3. Cap tubes, run PCR reaction, and run small amount (2 uL) on a gel to check for bands.\n",
    "\n",
    "    3.1. If bands are present, continue to final steps.\n",
    "    \n",
    "    3.2. If not, either determine which step to start over from, or start over from before you made the generic MMX.\n",
    "\n",
    "\n",
    "\n",
    "## Final Steps\n",
    "\n",
    "1. Remove \"uL to Remove\" from the individual MMXs as necessary according to the mmx table.\n",
    "\n",
    "2. Add \"uL DNA to Add\" DNA from the sample to the individual MMX according to the mmx table.\n",
    "\n",
    "3. Aliquot 100 uL per well to $n - 1$ wells to perform the remaining sample PCRs. (No additional neg. control necessary.)\n",
    "\n",
    "4. Run PCRs.\n",
    "\n",
    "5. Label plate on side with notched corner in format: \" $[DATE]\\ \\ \\ \\ SRM: [SRM Order Number]\\ \\ \\ \\ CRISPR$ \"\n",
    "\n",
    "    5.1. **Use Sharpie!** Hartwell specifically requested Sharpie rather than adhesive label to avoid lost labels."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Appendix\n",
    "---\n",
    "\n",
    "## Appendix A -- \"uL to Remove\"\n",
    "---\n",
    "\n",
    "### Description\n",
    "\n",
    "This value represents the value of uL that should be removed from the individual master mix before final aliquots. That is to say, before adding DNA to the master mix and subsequent aliquoting of the remaining PCR reactions required to have appropriate representation of the sample.\n",
    "\n",
    "Put simply, this value should be as follows:\n",
    "\n",
    "> $[uL \\ \\ to \\ \\ Remove] = [Rxns \\ \\ to \\ \\ remove] * [uL \\ \\ of \\ \\ MMX/rxn]$\n",
    "\n",
    "### Variable A: Reactions to Remove\n",
    "\n",
    "#### Equation\n",
    "\n",
    "$[Rxns \\ \\ to \\ \\ remove] = [current \\ \\ rxns \\ \\ in \\ \\ MMX] - [desired \\ \\ rxns]$\n",
    "\n",
    "#### Current Reactions in MMX:\n",
    "\n",
    "> $[current \\ \\ rxns] = [n\\_total] - [number \\ \\ of \\ \\ test \\ \\ rxns]$\n",
    "\n",
    "> $\\ => [current \\ \\ rxns] = n\\_total - 2 (neg. \\ \\ control \\ \\ and \\ \\ sample)$\n",
    "\n",
    "#### Desired Reactions:\n",
    "\n",
    "> $[desired \\ \\ rxns] = [remaining \\ \\ necessary \\ \\ PCRs] + [desired \\ \\ excess]$\n",
    "\n",
    "> $\\ => [desired \\ \\ rxns] = (n - 1) + 1 = n$\n",
    "\n",
    "> $\\ => [desired \\ \\ rxns] = n$\n",
    "\n",
    "#### Simplifying Rxns to Remove:\n",
    "\n",
    "> $[rxns \\ \\ to \\ \\ remove] = [current \\ \\ rxns] - [desired \\ \\ rxns]$\n",
    "\n",
    "> $\\ => (n\\_total - 2) - (n) = n\\_total - 2 - n$\n",
    "\n",
    "Because diff b/t n_total and n = EXTRA_RXNS, $(n\\_total - 2 - n)$ can be simplified to...\n",
    "\n",
    "> $[rxns \\ \\ to \\ \\ remove] = EXTRA\\_RXNS - 2$\n",
    "\n",
    "### Variable B: uL of MMX Per Reaction\n",
    "\n",
    "Because individual MMX now contains all components except for DNA...\n",
    "\n",
    "> $[uL DNA/rxn \\ \\ in \\ \\ MMX] = [rxn \\ \\ volume] - [missing \\ \\ volume \\ \\ of \\ \\ DNA]$\n",
    "\n",
    "> $\\ => [uL DNA/rxn \\ \\ in \\ \\ MMX] = 100 - [uL \\ \\ DNA \\ \\ to \\ \\ Add]$\n",
    "\n",
    "### SUMMARY\n",
    "\n",
    "> $[uL \\ \\ to \\ \\ Remove] = [rxns \\ \\ to \\ \\ remove] * [uL DNA/rxn \\ \\ in \\ \\ MMX]$\n",
    "\n",
    "> $[rxns \\ \\ to \\ \\ remove] = EXTRA\\_RXNS - 2$\n",
    "\n",
    "> $[uL DNA/rxn \\ \\ in \\ \\ MMX] = 100 - [uL \\ \\ DNA/rxn]$\n",
    "\n",
    "Therefore:\n",
    "\n",
    "> $[uL \\ \\ to \\ \\ Remove] = (EXTRA\\_RXNS - 2) * (100 - [uL \\ \\ DNA/rxn])$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cage37",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
